File: prim.cs
defaultEnv = {}

if window?
  # running in a browser
  window.global = window
  output = null
  defaultEnv.write = (msg)->
    if !output? then output = document.getElementById('output')
    output.innerHTML += "<span>#{msg}</span>"
    output.lastChild.scrollIntoView()
  defaultEnv.prompt = (msg, cont)-> cont(window.prompt(msg))
  window.Prim = root = {}
  Leisure = window.Leisure
else
  # running in node
  root = exports ? this
  Leisure = require('./leisure')
  Pretty = require('./pretty')
  U = require('util')
  RL = require('readline')
  tty = null
  defaultEnv.write = (msg)-> process.stdout.write(msg)
  defaultEnv.prompt = (msg, cont)-> tty.question(msg, cont)

setTty = (rl)-> tty = rl

define = Leisure.define
getType = Leisure.getType
laz = Leisure.laz

define 'is', (value)-> (type)-> if value()?.type == type().dataType then `_true()` else `_false()`
define 'eq', (a)-> (b)-> if a() == b() then `_true()` else` _false()`
define 'getType', (value)-> if type = getType(value()) then _some()(->type) else _none()
define 'parse', (value)->
  [prepped, err] = Leisure.prepare String(value())
  if err? then _right()(laz("Error: #{err}"))
  else
    [ast, err, rest] = Leisure.parseFull(prepped)
    if err? then _right()(laz("Error: #{err}"))
    else if rest?.trim() then _right()(laz("Error, input left after parsing: '#{rest.trim()}'"))
    else _left()(laz(ast))
define 'pretty', (value)-> Pretty.print(value())

define '+', (a)->(b)->a() + b()
define '-', (a)->(b)->a() - b()
define '*', (a)->(b)->a() * b()
define '/', (a)->(b)->a() / b()
define '%', (a)->(b)->a() % b()
define 'floor', (a)->Math.floor(a())
define 'ceil', (a)->Math.ceil(a())
define 'round', (a)->Math.round(a())

define 'randInt', (from)->(to)->
  makeMonad 'end', (env, cont)->
    cont Math.floor(from() + Math.random() * (to() - from() + 1))

define 'rand', ()->
  makeMonad 'end', (env, cont)->
    cont (Math.random())

define '>', (a)->(b)->if a() > b() then `_true()` else `_false()`
define '<', (a)->(b)->if a() < b() then `_true()` else `_false()`
define 'gt', (a)->(b)->if a() > b() then `_true()` else `_false()`
define 'gte', (a)->(b)->if a() >= b() then `_true()` else `_false()`
define 'lt', (a)->(b)->if a() < b() then `_true()` else `_false()`
define 'lte', (a)->(b)->if a() <= b() then `_true()` else `_false()`

eventCmds = []
running = false

leisureEvent = (leisureFuncName, evt)->
  currentEvent = evt
  monad = Leisure.eval("#{Leisure.nameSub(leisureFuncName)}()")(laz(evt))
  runMonad monad, defaultEnv, ->

runMonad = (monad, env, cont)->
  eventCmds.push ->
    runMonads monad, env, (value)->
      if eventCmds.length then eventCmds.shift()()
      running = false
      cont(value)
  if !running and eventCmds.length
    running = true
    eventCmds.shift()()

runMonads = (monad, env, cont)->
  if monad?.cmd?
    monad.cmd env, (value) ->
      if monad.binding? then runMonads monad.binding(-> value), env, cont
      else cont(value)
  else throw new Error("Attempted to run something that's not a monad")

# Make a new function and hide func and binding in properties on it
# making them inaccessible to pure Leisure code
# so people won't accidentally fire off side effects
makeMonad = (binding, guts)->
  m = ->
  m.cmd = guts
  m.type = 'monad'
  if binding != "end" then m.binding = binding
  m

define 'eventX', (evt)-> evt().x

define 'eventY', (evt)-> evt().y

define 'eventTargetId', (evt)-> evt().target.id

define 'return', (v)->
  makeMonad 'end', (env, cont)->cont(v())

define 'log', (msg)->(value)->
  if (msg().type != 'cons') then defaultEnv.write("#{msg()}") else defaultEnv.write(concatList(msg()))
  defaultEnv.write("\n")
  value()

define 'print', (msg)->
  makeMonad 'end', (env, cont)->
    if msg() != _nil() then env.write("#{msg()}\n")
    cont(_false)

define 'prompt', (msg)->
  makeMonad 'end', (env, cont)->
    env.prompt(String(msg()), (input)-> cont(input))

define 'bind', (m)->(binding)->
  makeMonad binding(), (env, cont)-> runMonads m(), env, cont

head = (l)->l ->(hh)->(tt)->hh()
tail = (l)->l ->(hh)->(tt)->tt()

concatList = (l)->
  if l == _nil() then ""
  else (head l) + concatList tail l

define 'concat', (l)-> concatList(l())

define 'js', (codeList)->
  makeMonad 'end', (env, cont)->
    cl = codeList()
    if cl != _nil() && cl.type != 'cons' then throw new Error("js expects a list for its code")
    cont(eval(concatList(cl)))

define 'browser', (codeList)->
  makeMonad 'end', (env, cont)->
    if window?
      cl = codeList()
      if cl != _nil() && cl.type != 'cons' then throw new Error("js expects a list for its code")
      cont(eval(concatList(cl)))
    else cont(null)

values = {}

define 'getValue', (name)->
  makeMonad 'end', (env, cont)->
    cont values[name()]

define 'setValue', (name)->(value)->
  makeMonad 'end', (env, cont)->
    values[name()] = value()
    cont _false

define 'createS', ->
  makeMonad 'end', (env, cont)->
    cont {value: null}

define 'getS', (state)->
  makeMonad 'end', (env, cont)->
    cont state().value

define 'setS', (state)->(value)->
  makeMonad 'end', (env, cont)->
    state().value = value()
    cont(_false)

root.setTty = setTty
root.runMonad = runMonad
root.makeMonad = makeMonad
root.tokenDefs = []
root.leisureEvent = leisureEvent
root.defaultEnv = defaultEnv

if window? then window.leisureEvent = leisureEvent

File: prim.js
(function() {
  var Leisure, Pretty, RL, U, concatList, defaultEnv, define, eventCmds, getType, head, laz, leisureEvent, makeMonad, output, r, root, runMonad, runMonads, running, setTty, tail, tty, values;

  defaultEnv = {};

  if (typeof window !== "undefined" && window !== null) {
    window.global = window;
    output = null;
    defaultEnv.write = function write(msg) {
      if (!(output != null)) output = document.getElementById('output');
      if (output != null) {
        output.innerHTML += "<span>" + msg + "</span>";
        return output.lastChild.scrollIntoView();
      }
    };
    defaultEnv.prompt = function prompt(msg, cont) {
      return cont(window.prompt(msg));
    };
    window.Prim = root = {};
    Leisure = window.Leisure;
  } else {
    root = typeof exports !== "undefined" && exports !== null ? exports : this;
    Leisure = require('./leisure');
    Pretty = require('./pretty');
    U = require('util');
    RL = require('readline');
    tty = null;
    defaultEnv.write = function write(msg) {
      return process.stdout.write(msg);
    };
    defaultEnv.prompt = function prompt(msg, cont) {
      return tty.question(msg, cont);
    };
    r = function r(file, cont) {
      console.log(U.inspect(file));
      if (!(file.match(/^\.\//))) file = "./" + file;
      require(file);
      return cont();
    };
    defaultEnv.require = r;
  }

  setTty = function setTty(rl) {
    return tty = rl;
  };

  define = Leisure.define;

  getType = Leisure.getType;

  laz = Leisure.laz;

  define('is', function(value) {
    return function(type) {
      var _ref;
      if (((_ref = value()) != null ? _ref.type : void 0) === type().dataType) {
        return _true();
      } else {
        return _false();
      }
    };
  });

  define('eq', function(a) {
    return function(b) {
      if (a() === b()) {
        return _true();
      } else {
        return  _false();
      }
    };
  });

  define('getType', function(value) {
    var type;
    if (type = getType(value())) {
      return _some()(function() {
        return type;
      });
    } else {
      return _none();
    }
  });

  define('parse', function(value) {
    var ast, err, prepped, rest, _ref, _ref2;
    _ref = Leisure.prepare(String(value())), prepped = _ref[0], err = _ref[1];
    if (err != null) {
      return _right()(laz("Error: " + err));
    } else {
      _ref2 = Leisure.parseFull(prepped), ast = _ref2[0], err = _ref2[1], rest = _ref2[2];
      if (err != null) {
        return _right()(laz("Error: " + err));
      } else if (rest != null ? rest.trim() : void 0) {
        return _right()(laz("Error, input left after parsing: '" + (rest.trim()) + "'"));
      } else {
        return _left()(laz(ast));
      }
    }
  });

  define('pretty', function(value) {
    return Pretty.print(value());
  });

  define('+', function(a) {
    return function(b) {
      return a() + b();
    };
  });

  define('-', function(a) {
    return function(b) {
      return a() - b();
    };
  });

  define('*', function(a) {
    return function(b) {
      return a() * b();
    };
  });

  define('/', function(a) {
    return function(b) {
      return a() / b();
    };
  });

  define('%', function(a) {
    return function(b) {
      return a() % b();
    };
  });

  define('floor', function(a) {
    return Math.floor(a());
  });

  define('ceil', function(a) {
    return Math.ceil(a());
  });

  define('round', function(a) {
    return Math.round(a());
  });

  define('randInt', function(from) {
    return function(to) {
      return makeMonad('end', function(env, cont) {
        return cont(Math.floor(from() + Math.random() * (to() - from() + 1)));
      });
    };
  });

  define('rand', function() {
    return makeMonad('end', function(env, cont) {
      return cont(Math.random());
    });
  });

  define('>', function(a) {
    return function(b) {
      if (a() > b()) {
        return _true();
      } else {
        return _false();
      }
    };
  });

  define('<', function(a) {
    return function(b) {
      if (a() < b()) {
        return _true();
      } else {
        return _false();
      }
    };
  });

  define('gt', function(a) {
    return function(b) {
      if (a() > b()) {
        return _true();
      } else {
        return _false();
      }
    };
  });

  define('gte', function(a) {
    return function(b) {
      if (a() >= b()) {
        return _true();
      } else {
        return _false();
      }
    };
  });

  define('lt', function(a) {
    return function(b) {
      if (a() < b()) {
        return _true();
      } else {
        return _false();
      }
    };
  });

  define('lte', function(a) {
    return function(b) {
      if (a() <= b()) {
        return _true();
      } else {
        return _false();
      }
    };
  });

  eventCmds = [];

  running = false;

  define('log', function(msg) {
    return function(value) {
      if (msg().type !== 'cons') {
        defaultEnv.write("" + (msg()));
      } else {
        defaultEnv.write(concatList(msg()));
      }
      defaultEnv.write("\n");
      return value();
    };
  });

  leisureEvent = function leisureEvent(leisureFuncName, evt, env) {
    var currentEvent, monad;
    currentEvent = evt;
    monad = Leisure.eval("" + (Leisure.nameSub(leisureFuncName)) + "()")(laz(evt));
    return runMonad(monad, env != null ? env : defaultEnv, function() {});
  };

  runMonad = function runMonad(monad, env, cont) {
    eventCmds.push(function() {
      return runMonads(monad, env, function(value) {
        if (eventCmds.length) eventCmds.shift()();
        running = false;
        return cont(value);
      });
    });
    if (!running && eventCmds.length) {
      running = true;
      return eventCmds.shift()();
    }
  };

  runMonads = function runMonads(monad, env, cont) {
    if ((monad != null ? monad.cmd : void 0) != null) {
      return monad.cmd(env, function(value) {
        if (monad.binding != null) {
          return runMonads(monad.binding(function() {
            return value;
          }), env, cont);
        } else {
          return cont(value);
        }
      });
    } else {
      throw new Error("Attempted to run something that's not a monad");
    }
  };

  makeMonad = function makeMonad(binding, guts) {
    var m;
    m = function m() {};
    m.cmd = guts;
    m.type = 'monad';
    if (binding !== "end") m.binding = binding;
    return m;
  };

  define('eventContext', function(evt) {
    return evt().leisureContext;
  });

  define('eventX', function(evt) {
    return evt().x;
  });

  define('eventY', function(evt) {
    return evt().y;
  });

  define('eventTargetId', function(evt) {
    return evt().target.id;
  });

  define('return', function(v) {
    return makeMonad('end', function(env, cont) {
      return cont(v());
    });
  });

  define('require', function(file) {
    return makeMonad('end', function(env, cont) {
      return env.require(file(), cont);
    });
  });

  define('print', function(msg) {
    return makeMonad('end', function(env, cont) {
      if (msg() !== _nil()) env.write("" + (msg()) + "\n");
      return cont(_false());
    });
  });

  define('prompt', function(msg) {
    return makeMonad('end', function(env, cont) {
      return env.prompt(String(msg()), function(input) {
        return cont(input);
      });
    });
  });

  define('bind', function(m) {
    return function(binding) {
      return makeMonad(binding(), function(env, cont) {
        return runMonads(m(), env, cont);
      });
    };
  });

  head = function head(l) {
    return l(function() {
      return function(hh) {
        return function(tt) {
          return hh();
        };
      };
    });
  };

  tail = function tail(l) {
    return l(function() {
      return function(hh) {
        return function(tt) {
          return tt();
        };
      };
    });
  };

  concatList = function concatList(l) {
    if (l === _nil()) {
      return "";
    } else {
      return (head(l)) + concatList(tail(l));
    }
  };

  define('concat', function(l) {
    return concatList(l());
  });

  define('js', function(codeList) {
    return makeMonad('end', function(env, cont) {
      var cl;
      cl = codeList();
      if (cl !== _nil() && cl.type !== 'cons') {
        throw new Error("js expects a list for its code");
      }
      return cont(eval(concatList(cl)));
    });
  });

  define('browser', function(codeList) {
    return makeMonad('end', function(env, cont) {
      var cl;
      if (typeof window !== "undefined" && window !== null) {
        cl = codeList();
        if (cl !== _nil() && cl.type !== 'cons') {
          throw new Error("js expects a list for its code");
        }
        return cont(eval(concatList(cl)));
      } else {
        return cont(null);
      }
    });
  });

  values = {};

  define('getValue', function(name) {
    return makeMonad('end', function(env, cont) {
      return cont(values[name()]);
    });
  });

  define('setValue', function(name) {
    return function(value) {
      return makeMonad('end', function(env, cont) {
        values[name()] = value();
        return cont(_false);
      });
    };
  });

  define('createS', function() {
    return makeMonad('end', function(env, cont) {
      return cont({
        value: null
      });
    });
  });

  define('getS', function(state) {
    return makeMonad('end', function(env, cont) {
      return cont(state().value);
    });
  });

  define('setS', function(state) {
    return function(value) {
      return makeMonad('end', function(env, cont) {
        state().value = value();
        return cont(_false);
      });
    };
  });

  root.setTty = setTty;

  root.runMonad = runMonad;

  root.makeMonad = makeMonad;

  root.tokenDefs = [];

  root.leisureEvent = leisureEvent;

  root.defaultEnv = defaultEnv;

  if (typeof window !== "undefined" && window !== null) {
    window.leisureEvent = leisureEvent;
  }

}).call(this);
